### PR TITLE
Modularized three main charts

### PR DESCRIPTION
**Pull request by CBarrett2**
This isn't to pull quite yet, but spent some time splitting the charts up into different files. I've used events to communicate changes in view and when to render. Thoughts?

### CODE REVIEW

**Comment by djmiley** [comment]
#70

**Comment by CBarrett2** [comment]
Resolves #115 #116 and #117 plus sets up an events system that should make some other modularizations easier

**Comment by CBarrett2** [comment]
Would appreciate another look-over of this :)

File: src/assets/js/sc.js [hunk 1]
```diff
@@ -13,6 +13,7 @@
                 coinbase: {}
             }
         },
-        util: {}
+        util: {},
+        chart: {}
     };
```
Comment (djmiley) [inline src/assets/js/sc.js]: Alphabetical order?

File: src/assets/js/main.js [hunk 2]
```diff
@@ -61,75 +69,25 @@
                 .node()
                 .value;
             changeSeries(seriesTypeString);
+            render();
         });
```
Comment (djmiley) [inline src/assets/js/main.js]: Is there a reason you've changed `changeSeries`?

File: src/assets/js/main.js [hunk 3]
```diff
@@ -12,23 +12,30 @@
     var ohlc = fc.series.ohlc();
     var point = fc.series.point();
     var line = fc.series.line();
+    line.isLine = true;
```
Comment (djmiley) [inline src/assets/js/main.js]: My menu request should solve this hopefully.

File: src/assets/js/main.js (hunk 2 above)
Comment (CBarrett2) [inline src/assets/js/main.js]: It no longer has access to the multi inside primaryChart, so it has to go through primaryChart.changeSeries now.

File: src/assets/js/main.js (hunk 3 above)
Comment (CBarrett2) [inline src/assets/js/main.js]: I saw your solution to that, I like it!

File: src/assets/js/main.js (hunk 2 above)
Comment (djmiley) [inline src/assets/js/main.js]: Yes, that's good.

File: src/assets/js/chart/navChart.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                return data;
+            });
+
+            navTimeSeries.plotArea(navMulti);
+            selection.call(navTimeSeries);
+        }
+
+        navChart.onViewChange = function(func) {
+            dispatch.on('viewChange', func);
+        };
```
Comment (djmiley) [inline src/assets/js/chart/navChart.js]: `viewChange.navChart` ?

File: src/assets/js/chart/primaryChart.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            dispatch.on('viewChange.primaryChart', func);
+        };
+
+        primaryChart.changeSeries = function(series) {
+            multi.series([gridlines, movingAverageLine, series, closeAxisAnnotation]);
+            return primaryChart;
+        };
+
+        d3.rebind(primaryChart, timeSeries, 'xDomain');
+
```
Comment (djmiley) [inline src/assets/js/chart/primaryChart.js]: Is this needed? Have a check.

**Comment by djmiley** [comment]
'viewChange' -> 'viewChanged'? Other than that, looks good.

File: src/assets/js/chart/primaryChart.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            });
+
+            // Scale y axis
+            var yExtent = fc.util.extent(sc.util.filterDataInDateRange(data, timeSeries.xDomain()), ['low', 'high']);
+            // Add 10% either side of extreme high/lows
+            var variance = yExtent[1] - yExtent[0];
+            yExtent[0] -= variance * 0.1;
+            yExtent[1] += variance * 0.1;
+            timeSeries.yDomain(yExtent);
+
```
Comment (djmiley) [inline src/assets/js/chart/primaryChart.js]: Also remove lines 88 to 91. #91 #109

File: src/assets/js/chart/navChart.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                }
+                return data;
+            });
+
+            navTimeSeries.plotArea(navMulti);
+            selection.call(navTimeSeries);
+        }
+
+        navChart.onViewChange = function(func) {
+            dispatch.on('viewChange.navChart', func);
```
Comment (tlsim) [inline src/assets/js/chart/navChart.js]: Does the namespace make any difference here?

Comment (CBarrett2) [inline src/assets/js/chart/navChart.js]: I don't think so

File: src/assets/js/main.js [hunk 8]
```diff
@@ -12,23 +12,30 @@
...
+    var navChart = sc.chart.navChart();
+
+    function onViewChange(domain) {
+        data.viewDomain = [domain[0], domain[1]];
+        render();
+    }
 
-    var standardDateDisplay = [data[Math.floor((1 - navAspect * goldenRatio) * data.length)].date,
-        data[data.length - 1].date];
+    primaryChart.onViewChange(onViewChange);
```
Comment (tlsim) [inline src/assets/js/main.js]: prefer `primaryChart.on('viewChange' onViewChanged)` so it's clear that it's using d3 events.

File: src/assets/js/main.js [hunk 9]
```diff
@@ -12,23 +12,30 @@
...
+    line.isLine = true;
     var area = fc.series.area();
 
-    var currentSeries = candlestick;
-
     var data = fc.data.random.financial()(250);
 
-    // Using golden ratio to make initial display area rectangle into the golden rectangle
-    var goldenRatio = 1.618;
+    data.viewDomain = [];
```
Comment (tlsim) [inline src/assets/js/main.js]: Would be better not to store this on the data array itself. Bind a model with both `data` and `viewDomain` instead.

**Comment by tlsim** [comment]
Definitely moving in the right direction. Seeing it split up certainly makes things easier to follow! I'm wondering if rather than each chart creating a new 'viewScale' and using events to communicate the domain change, they could instead share one instance of an xScale? It's also surprising to see 3 d3.behavior.zoom components, but we can work on that later in #121.

**Comment by CBarrett2** [comment]
viewScale is only created in navChart, and I think that requires the range being navChart's width, as opposed to the width of any other chart. The rest just use the domain passed to them. It's possible for all of them to share the same scale, but I thought we agreed to do it this way?

**Comment by tlsim** [comment]
@CBarrett2 Ok we can think about if shared scales make any sense later.

**Pull request merged**